### ISSUE TITLE
Google Authenticator 2FA users: I NEED YOU

### ISSUE THREAD

**Issue by snipe** [snipe/snipe-it#2847]
Hey gang,

I've just pushed some changes up to `develop` that finally enabled Google Authenticator two-factor authentication.

<img width="434" alt="screen shot 2016-10-29 at 10 18 46 am" src="https://cloud.githubusercontent.com/assets/197404/19832683/6516d026-9ddf-11e6-9981-2743354b891a.png">

There are thee different settings in the Admin > Settings:
- 2FA disabled
- Optional (a user can choose to enable it for themselves)
- Required (ALL users must use 2FA)
### Workflow

If you try to login when the 2FA settings are set to Optional and your account has 2FA enabled, OR if the 2FA settings are set to required, you should be prompted one-time to enroll your device after you login. You'll be presented with a QR code to add the configuration to your device, and you should see it pop up right away in your Google Auth app.

After that, every time you login, you should be promoted for login, and then your Auth code on the next page.
### How to Help Test

First of all, DO NOT USE YOUR PRODUCTION DATABASE FOR THIS. While I don't think anything we've done should cause breakage, it's just a good idea to duplicate your production database on a development box and use that. You'll need to pull from the `develop` branch instead of master, or an official release.

Looking forward to hearing your feedback!

**Comment by delinski111** [comment]
On my home PC. This worked great. Got two accounts registered on a single device, just scanned, put in the code, and it just worked.

On a dev VM at my work, I tried this. Enabled the optional 2FA option in settings. scanned the QR, got the device to start sending codes, but the webpage will not accept any of them. I also now cannot disable 2FA on my account and just cannot get into anymore. (I have other admin accounts, but selecting reset device, and deactivating the profile will not stop it from prompting for 2FA).

**Comment by snipe** [comment]
Huh. I have no idea why you'd see different behavior between those two scenarios. Is it actually telling you that the code is wrong, or just bouncing you back to the code page?

Which 2FA requirement did you pick, optional or required?

If you have access to the DB, update the `two_factor_enabled` field to NULL, if you're locked out.

**Comment by snipe** [comment]
Does this behavior change on your VM if you use different browsers?

**Comment by base-zero** [comment]
Make sure that the time on the VM is set correctly too.

**Comment by snipe** [comment]
@base-zero EXCELLENT catch there. That can definitely cause problems (and I know the time on my VMs is usually wrong.)

**Comment by base-zero** [comment]
👍  I spent a few hours debugging the same thing on one of my Laravel apps and in the end the code was fine and it was all down to the VM's time. ```ntpdate ntp.ubuntu.com``` is a keeper :)

**Comment by delinski111** [comment]
Error:  Invalid two-factor code 
"Optional" is the selection I am currently testing.
Using a different browser resulted in same issue.

Good Call about the timezone. It is incorrect.
If you are willing to help me troubleshoot basic linux.
I ran sudo timedatectl set-timezone EST - Time was not correct still (it kept default to EDT?)
I ran the command you suggested, no server suitable for sync found.
I ran dpkg-reconfigure tzdata and selected my region, still incorrect. not really sure what up here

**Comment by delinski111** [comment]
So, I now selected "Disabled" for the 2FA option, and it kicked out my admin user (who did not have 2FA enabled in the first place) and now it is requireing 2FA from this user, even though the option is now set to disabled.

**Comment by snipe** [comment]
I don't have any idea why that would be happening. This is the entire middleware:

```
// Skip the logic if the user is on the two factor pages
        if (($request->route()->getName()=='two-factor') || ($request->route()->getName()=='two-factor-enroll') || ($request->route()->getName()=='logout')) {
            return $next($request);
        }

        // Two-factor is enabled (either optional or required)
        if (Auth::check() && (Setting::getSettings()->two_factor_enabled!='')) {

            // This user is already 2fa-authed
            if ($request->session()->get('2fa_authed')){
                return $next($request);
            }

            // Two-factor is optional and the user has NOT opted in, let them through
            if ((Setting::getSettings()->two_factor_enabled=='1') && (Auth::user()->two_factor_optin!='1')) {
                return $next($request);
            }

            // Otherwise make sure they're enrolled and show them the 2FA code screen
            if ((Auth::user()->two_factor_secret!='') && (Auth::user()->two_factor_enrolled=='1')) {
                return redirect()->route('two-factor')->with('info', 'Please enter your two-factor authentication code.');
            }

            return redirect()->route('two-factor-enroll')->with('success', 'Please enroll a device in two-factor authentication.');
            

        }

        return $next($request);
```

**Comment by delinski111** [comment]
I am now back at my home machine and will keep messing with things. If there are any steps or screenshots you would like me to get/attempt let me know and I will send you what I can.

To be clear, now that I am looking at my original post. I checked out the "features/drawn_signatures" branch, and am testing that. Is there a difference between develop / drawn_signatures in terms of the 2FA feature (maybe I am missing some file)?

**Comment by snipe** [comment]
I would switch back to the dev branch to test this. There *shouldn't* be differences, but it's probably better to use the right branch.

**Comment by delinski111** [comment]
I manually set 2FA - null in backend. It worked, got into server.

Navigated to Settings, clicked edit, verified that 2FA was disabled. Clicked save (I did not enable it) and it kicked me out and reenabled automatically. (still did not change branches yet).

![2fa_hates_me](https://cloud.githubusercontent.com/assets/22087997/19870407/cbbbdb48-9f86-11e6-8367-97b70413d0a1.gif)

**Comment by snipe** [comment]
Can you tell me what the value is in MySQL for `two_factor_enabled`? I may have found the issue, but would like to confirm from your side.

**Comment by delinski111** [comment]
![image](https://cloud.githubusercontent.com/assets/22087997/19870620/bc086396-9f87-11e6-9919-216f0a4c30cf.png)


![image](https://cloud.githubusercontent.com/assets/22087997/19870594/99c56914-9f87-11e6-9ca9-81f51ab82a0a.png)

**Comment by snipe** [comment]
Crap. Then no, that's not it. In my testing, I'm seeing it set the value of two_factor_enabled to 0 instead of null when it saves. Was hoping that might have been it, but it's showing as null for you in your DB. :-/

**Comment by snipe** [comment]
Wait, is that the structure or the actual value you're showing me? I'm looking for the value, once you disable 2FA through the settings admin and save

**Comment by delinski111** [comment]
I manually told it to do that after not being able to get back in.
Just tried UPDATE settings SET two_factor_enabled = 0;
but that did not actually disable it, needs to be set to NULL
![image](https://cloud.githubusercontent.com/assets/22087997/19870758/5cea657a-9f88-11e6-99f1-e1018442b5b8.png)

![image](https://cloud.githubusercontent.com/assets/22087997/19870778/6fdc4360-9f88-11e6-8ce8-61fa174830b9.png)

After making NULL. I can get in. Once I click settings, and save, it reverts back to 0. That is the problem, whatever is triggering that value to change.

**Comment by snipe** [comment]
Can you grab the latest from develop and see if you still have the problem?

**Comment by delinski111** [comment]
Hmm, I have a runtime error for no supported encrypter. I re altered ~/config/app.php to say "AES-256-CBC" and restarted apache. Error still there. Key is the same (nothing was encrypted anyway as this is a test db with 1 assett)

**Comment by snipe** [comment]
WTF. I didn't change anything that should affect your encryption. Does your app key start with `base64:`?

**Comment by snipe** [comment]
(Also try regenerating the app key `php artisan key:generate` - it shouldn't matter, since this is a test.)

**Comment by delinski111** [comment]
Yes.

Regenerating did not change anything

I can just blow out this build. I have reinstalled so many server of this while testing i can  get another one up in about 6ish minutes if you want me to go that route.

**Comment by snipe** [comment]
If you don't mind, yeah. Your situation is a little different, since you opted to use the AES stuff ahead of when we were ready (so you didn't have to use mcrypt). That's an additional variable that I think could be conflating things, as there is literally no change I made that should have impacted your ciphers.

**Comment by delinski111** [comment]
Same issue, fresh server, fresh checkout, edit settings, change nothing, click save, brings me to 2fa QR screen

**Comment by snipe** [comment]
Same questions then - whats' the value of `two_factor_enabled`? (And you did a fresh pull from develop, right, not the drawn signatures branch)?

**Comment by delinski111** [comment]
Yes develop this time :D 

Value equals 0 again.

**Comment by snipe** [comment]
Can you show me line 321 - 325 in app/Http/Controllers/SettingsController.php?

**Comment by delinski111** [comment]
![image](https://cloud.githubusercontent.com/assets/22087997/19871865/2dfb225e-9f8d-11e6-8cb9-f4348e667e33.png)

**Comment by snipe** [comment]
Yeah, that's the older code, you didn't pick up what I just pushed out: https://github.com/snipe/snipe-it/commit/e065d18227186757ad22f138cbabc27a4a43dbe3

**Comment by delinski111** [comment]
I just checked it out. odd. What is the solution? do i do git pull and it will pull from the branch I am using?

**Comment by snipe** [comment]
Yes, git pull

**Comment by delinski111** [comment]
welp, yes things were changed.

That seems like it worked. Great. Thank you for taking the time on this one. I dont really understand how the checkout process missed those changes. I guess from now on it would be best practice for me to just always try a pull before digging deep into troubleshooting.

**Comment by delinski111** [comment]
Atleast we know that the AES method isnt broken 👍

**Comment by snipe** [comment]
Awesome - thank YOU so much for helping us test this :) It's a pretty big feature and we're excited about it (and tested it locally, obviously), but we really appreciate all the time you've taken to help out. If we had pushed out that bug, that would have been a real mess.

**Comment by delinski111** [comment]
I am not proficient in programming, so the best I can do is be the Guinea pig, especially because you make this free for me to use you glorious person :D!

**Comment by snipe** [comment]
> I am not proficient in programming 

LOL based on the number of bugs I write, apparently I'm not either ;)

> so the best I can do is be the Guinea pig, 

It helps SO much, I can't even tell you. Especially in an app like this where there are so many possible ways of using it, different configurations, different versions, etc. It's always a bug help.

> especially because you make this free for me to use you glorious person :D!

Awww... thanks! <3

![687474703a2f2f7777772e7265616374696f6e676966732e636f6d2f722f746366692e676966](https://cloud.githubusercontent.com/assets/197404/19872192/b71efa82-9f75-11e6-8ebc-e901dfcc78bf.gif)

**Comment by delinski111** [comment]
Using the same server as above. A new concern is now alive.

I am not being kicked out anymore at all for 2FA. If i manually log out, and log back in, it works fine (yay). It will navigate over to the device registration or code field if previously registered.

**Comment by snipe** [comment]
I don't understand - can you clarify that a little?

**Comment by delinski111** [comment]
That was way too fast. like, 12 seconds after i made the post. It started to work immediately after submitting that ticket. (I deleted the original comment but you beat me to it) (original comment, for the record was that I was enabling 2FA on a user profile was not kicking me out to register a device)

**Comment by therealjoshuad** [comment]
[Possible bug?] I was having trouble at first. I was getting an "invalid barcode" in the Google Auth app when trying to scan the barcode. I noticed in the URL that the barcode was generating contained the site name, which contained a space. I removed the space from the site name, and it worked. 

I Tested a couple of logins after that, and they all worked successfully.

After pairing the app, and putting the space back in the site title, it still works.
